### PR TITLE
Merkle tree tracker

### PR DESCRIPTION
**Pull request by MarekM25**

### CODE REVIEW

File: src/Nethermind/Nethermind.Baseline.Test/BaselineTreeTrackerReorganizationTests.cs [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+            for (int j = 0; j < test.LeavesBeforeReorganization.Length; j++)
+            {
+                var leafHash = test.LeavesBeforeReorganization[j];
+                var transaction = contract.InsertLeaf(address, leafHash);
+                transaction.Nonce = nonce;
+                ++nonce;
+                await testRpc.TxSender.SendTransaction(transaction, TxPool.TxHandlingOptions.None);
+            }
+
+            await testRpc.AddBlock(false);
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.Baseline.Test/BaselineTreeTrackerReorganizationTests.cs:74]: how is reorganization tested here?

File: src/Nethermind/Nethermind.Baseline/Tree/BaselineTree.cs [hunk 2]
```diff
@@ -19,26 +22,136 @@ public abstract partial class BaselineTree
...
+        }
+
+        public BatchWrite StartBatch()
+        {
+            return new BatchWrite(_synchroObject);
+        }
+
+        #region Metadata operation
+
+        private const long CurrentBlockIndex = -1;
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.Baseline/Tree/BaselineTree.cs:88]: seems a bit strange to keep the block info in the BaselineTree class - why not a separate class?

File: src/Nethermind/Nethermind.Baseline/Tree/BaselineTree.cs [hunk 3]
```diff
@@ -19,26 +22,136 @@ public abstract partial class BaselineTree
...
+        private void InitializeMetadata()
+        {
+            var currentBlock = LoadCurrentBlockInDb();
+            LastBlockDbHash = currentBlock.LastBlockDbHash;
+            LastBlockWithLeaves = currentBlock.LastBlockWithLeaves;
             Count = LoadCount();
-            Root = Keccak.Zero; // TODO: need to check what should be the initial root of an empty tree
+        }
+        public uint GetLeavesCountFromNextBlocks(long blockNumber, bool clearPreviousCounts = false)
+        {
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.Baseline/Tree/BaselineTree.cs:100]: I think this should all be in a separate class and not in the Tree itself

File: src/Nethermind/Nethermind.Baseline/Tree/BaselineTree.cs [hunk 4]
```diff
@@ -19,26 +22,136 @@ public abstract partial class BaselineTree
...
+            var rlpEncoded = _metadataKeyValueStore[MetadataBuildDbKey(CurrentBlockIndex)];
+            if (rlpEncoded == null)
+                return (ZeroHash, 0);
+            var rlpStream = new RlpStream(rlpEncoded);
+            rlpStream.SkipLength();
+            return (rlpStream.DecodeKeccak(), rlpStream.DecodeLong());
+        }
+
+        public void SaveCurrentBlockInDb()
+        {
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.Baseline/Tree/BaselineTree.cs:143]: Same - separate class IMHO

File: src/Nethermind/Nethermind.Blockchain/Filters/Topics/SequenceTopicsFilter.cs [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+//  You should have received a copy of the GNU Lesser General Public License
+//  along with the Nethermind. If not, see <http://www.gnu.org/licenses/>.
+
+using Nethermind.Blockchain.Receipts;
+using Nethermind.Core;
+using Nethermind.Core.Crypto;
+
+namespace Nethermind.Blockchain.Filters.Topics
+{
+    public class SequenceTopicsFilter : TopicsFilter
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.Blockchain/Filters/Topics/SequenceTopicsFilter.cs:23]: some comment would be nice to explain how it differs from the ANyTopicFilter

**Review by tkstanczak — COMMENTED** [review]
impressive

**Review by tkstanczak — APPROVED** [review]

File: src/Nethermind/Nethermind.Baseline/Tree/BaselineTreeTracker.cs [hunk 6]
```diff
@@ -13,53 +13,137 @@
...
+                {
+                    Reorganize(e.TxReceipts, e.Block.Number);
+                }
+
+                _currentBlockHeader = e.Block.Header;
+                return;
+            }
+
+            _currentBlockHeader = e.Block.Header;
+            AddFromCurrentBlock(e.TxReceipts, e.Block.Number);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Baseline/Tree/BaselineTreeTracker.cs:92]: when you reorganize, you can still AddFromCurrentBlock here, will make code easier, unless its hashing optimazation for now.

**Review by tkstanczak — APPROVED** [review]

File: src/Nethermind/Nethermind.Baseline/Tree/BaselineTree.cs [hunk 7]
```diff
@@ -46,8 +64,9 @@ public BaselineTree(IDb db, IKeyValueStore metadataKeyValueStore, byte[] _dbPref
             TruncationLength = truncationLength;
             _db = db ?? throw new ArgumentNullException(nameof(db));
             _metadataKeyValueStore = metadataKeyValueStore ?? throw new ArgumentNullException(nameof(metadataKeyValueStore));
-
             this._dbPrefix = _dbPrefix;
+            Metadata = new BaselineTreeMetadata(metadataKeyValueStore, _dbPrefix);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Baseline/Tree/BaselineTree.cs:68]: I think the idea was to decouple it and reverse dependencies. Not just putting it into another class.

Comment (tkstanczak) [reply src/Nethermind/Nethermind.Baseline/Tree/BaselineTree.cs:68]: yes, let us leave it as a technical debt for the days after the presentation

**Pull request merged**